design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,275 @@
+
+## HCO: Integration with SSP Operator
+
+In order to leverage golden disk image auto delivery/updates at the HCO level, the HCO is responsible for configuring DataImportCrons on the SSP operator’s CR.
+
+For instance, if the HCO wants to develop a feature for automatically sourcing in fedora golden images into a PVC to be used by the SSP's common templates, then the HCO needs to configure the location where the fedora image is sourced from on the SSP CR. 
+
+This document doesn't include within its scope how the HCO should expose enabling the image auto delivery/update mechanism on the HCO API or where the containerDisk images are sourced.
+
+## OCP UI: Integration with DataSources
+
+The UI currently detects if a template has a boot source by checking if the DataVolumeTemplate’s source PVC is present. The UI needs to now be capable of checking if a DataVolumeTemplate’s DataSource has the a ready condition equal to 'true'.
+
+## Scalability
+
+- DataImportCrons have the potential to exhaust storage resources. This can be mitigated through a garbage collection policy for out-of-date disk volumes. 
+- A large amount of DataImportCrons triggering at the exact same interval could impact network resources during the import. This can be mitigated by introducing a jitter factor into the recurring poll interval.
+
+## Update/Rollback Compatibility
+
+### Transitioning the CommonTemplates to DataSources
+
+SSP operator must make image pipelines an opt-in feature to ensure existing boot images are not overwritten.
+
+To handle the default (non auto update/delivery) use case during update, the SSP operator will create corresponding DataSources for every DataSource referenced in the DataVolumeTemplate sections present in the deployed common templates. These default DataSources will point to a PVC name that matches the PVC name used before DataSources were introduced into the common templates.
+
+Here's an example of how this works.
+
+**Before Update:** If the common templates had a DataVolumeTemplate that looked like the following before DataSources were introduced.
+```yaml
+dataVolumeTemplates:
+- apiVersion: cdi.kubevirt.io/v1beta1
+  kind: DataVolume
+  metadata:
+    name: fedora
+  spec:
+    pvc:
+      accessModes:
+        - ReadWriteMany
+      resources:
+        requests:
+          storage: 30Gi
+    source:
+      pvc:
+        name: fedora
+        namespace: golden-images
+```
+
+
+**After Update:** The common template would have a DataVolumeTemplate that looks like the following after the update
+```yaml
+dataVolumeTemplates:
+- apiVersion: cdi.kubevirt.io/v1beta1
+  kind: DataVolume
+  metadata:
+    name: fedora
+  spec:
+    pvc:
+      accessModes:
+        - ReadWriteMany
+      resources:
+        requests:
+          storage: 30Gi
+    dataSource:
```

Comment:
The embedding makes sense on a technical side, but not for the user (i think).
Everything you said sounds right, but none of the information is visible on the api (by naming etc.) which makes it implicit knowledge of the storage team. Not even the docs mention anything of this (http://kubevirt.io/api-reference/v0.40.0/definitions.html#_v1alpha1_datavolumesource) and I'd prefer to make the api more expressive instead of confusing.

So if `source` currently contains sources and those are mostly embedded (what doesn't seem true for pvc as it's also a ref), I don't see a reason why we shouldn't be able to do `.spec.source.reference` which explicitly adds an option to reference a data source that is not embedded here or part of this explicit api.
This reference field could even be made to in the future replace the pvc field so any kind of reference we come up in the future can go there. But I'm not fixed on the name.

That would make the most sense to me and fit the documentation of the `source` field.

Reply:
Another option I might be able to live with would be `.spec.sourceRef` as it's more in line with what is done in Kubernetes already and makes more sense as a replacement for the `source` field but referencing something else.
Just like a `secretRef`. But then again this is something I'd say is more and more discouraged from the K8s api side.

Reply:
I addition to what was said by @davidvossel, the choice to add an additional field was also influenced by the recent addition of the `spec.storage` field to DataVolumes.  See here:  https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/apis/core/v1beta1/types.go#L50

`spec.storage` basically has the same fields of `spec.pvc` but has fewer required fields and takes advantage of shared `StorageProfiles`:  https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/apis/core/v1beta1/types.go#L319-L325

A user can specify `spec.storage` or `spec.pvc` based on their needs.  Similar to proposed `spec.dataSource` and `spec.source`.  `spec.storage` use `spec.dataSource` take advantage of shared configuration resources.

I do like the idea of changing the name of `spec.dataSource` to `spec.sourceRef` though.

Reply:
Per our discussion. i documented using sourceRef like this.
```
apiVersion: cdi.kubevirt.io/v1beta1
kind: DataVolume
metadata:
   name: my-fedora-vm-image
spec:
  pvc:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: 2Gi
  sourceRef:
    kind: DataSource
    name: fedora
    namespace: golden-images
```

Reply:
I agree with @davidvossel  and @mhenriks having separate fields `dataVolume.Spec.Source` and `dataVolume.Spec.SourceRef` better represents these are different ways to define a source - directly or by reference to exteral object pointing to the actual source. We will sure update the documentation to make it clear for users.